File: src/examples/ex_table_pick_place.cpp
/* ex_table_pick_place.cpp
 * Kinova GEN3 + Robotiq 2F-85 on a table picks a cube and places it 0.24 m to the side, with
 * joint impedance (PD + KDL gravity) in TORQUE mode tracking IK waypoints. While the arm carries
 * the cube (PLACE_ABOVE), a scripted 20 N push across the motion on bracelet_link makes the arm
 * give way and spring back, as the joint stiffness Kp predicts.
 *
 * Usage:
 *   ex_table_pick_place [--headless]
 *
 * --headless skips the viewer and exits 1 unless the cube rests on the table within kMaxPlaceErr
 * of the place spot, the push deflected the TCP by at least kMinDeflection and within
 * kDeflectionRatio of J K^-1 J^T F, the arm sprang back, the cube stayed in the gripper and the
 * elbow stayed above kMinElbowHeight. */

#include "common.hpp"
#include "example_paths.hpp"

#include <kdl/chainjnttojacsolver.hpp>

#include <algorithm>
#include <cmath>
#include <iomanip>
#include <iostream>
#include <string>

namespace ex = mj_kdl_examples;

static constexpr double kMaxPlaceErr = 0.005; // [m] in the table plane

static constexpr const char *kPushBody  = "bracelet_link";
static constexpr const char *kPushPhase = "PLACE_ABOVE";
static constexpr double      kPushForce = 20.0; // [N] along world +x, across the carry along +y
static constexpr double      kPushOn    = 0.8;  // [s] into the phase
static constexpr double      kPushOff   = 2.0;  // [s] into the phase
static constexpr double      kPushRamp  = 0.2;  // [s] rise and fall
static constexpr double      kSettle    = 0.5;  // [s] after the push, when the residual is taken

static constexpr double kMinDeflection     = 0.010; // [m]
static constexpr double kDeflectionRatio[] = { 0.85, 1.15 }; // measured / predicted
static constexpr double kMaxResidual       = 0.004;          // [m]
static constexpr double kMaxCubeSlip       = 0.010; // [m] cube centre from the TCP while pushed
static constexpr double kMinElbowHeight    = 0.45;  // [m] forearm_link origin above the table

struct Metrics
{
    KDL::Vector lag0;                 // TCP minus its reference just before the push
    double      peak      = 0.0;      // [m] along the push
    double      predicted = 0.0;      // [m] J K^-1 J^T F at the peak
    double      residual  = -1.0;     // [m] along the push, kSettle after it; < 0 until taken
    double      cube_slip = 0.0;      // [m]
    double      elbow_min = INFINITY; // [m] world z
};

static double push_scale(double t_rel)
{
    return std::min(
      ex::clamp01((t_rel - kPushOn) / kPushRamp), ex::clamp01((kPushOff - t_rel) / kPushRamp)
    );
}

int main(int argc, char *argv[])
{
    const bool headless = ex::parse_args(argc, argv).headless;

    mj_kdl::RobotSpec robot_spec;
    robot_spec.path   = ex::menagerie_model("kinova_gen3/gen3.xml");
    robot_spec.pos[2] = ex::kTableZ;
    robot_spec.attachments.push_back(ex::gripper_attachment(ex::asset("robotiq_2f85/2f85.xml")));

    mj_kdl::SceneSpec scene = ex::scene_spec();
    scene.robots.push_back(robot_spec);
    scene.objects.push_back(ex::table_object(ex::asset("table.xml"), ex::kTableZ));
    scene.objects.push_back(ex::cube_object(ex::kPickXY[0], ex::kPickXY[1], ex::kTableZ));

    mj_kdl::Env   env;
    mj_kdl::Robot robot;
    if (!mj_kdl::init_env(&env, &scene)) return 1;
    const mj_kdl::ToolFrameSpec tool = ex::gripper_tool();
    if (!mj_kdl::init_robot_from_mjcf(&robot, &env, "base_link", "bracelet_link", "", &tool))
        return 1;
    if (!mj_kdl::set_control_mode(&robot, mj_kdl::CtrlMode::TORQUE)) return 1;
    mj_kdl::SceneActuatorSlot *fingers =
      mj_kdl::bind_scene_actuator(&env.scene, "g_fingers_actuator");
    mj_kdl::SceneFreeBodySlot *cube = mj_kdl::bind_scene_free_body(&env.scene, "cube");
    mj_kdl::SceneWrenchSlot   *push = mj_kdl::bind_scene_wrench(&env.scene, kPushBody);
    if (!fingers || !cube || !push) return 1;

    ex::PickPlaceWaypoints wp;
    if (!ex::solve_pick_place(robot, wp)) return 1;
    const unsigned                  n = robot.chain.getNrOfJoints();
    KDL::ChainDynParam              dyn(robot.chain, KDL::Vector(0.0, 0.0, scene.gravity_z));
    KDL::ChainFkSolverPos_recursive fk(robot.chain);
    KDL::ChainJntToJacSolver        jac_solver(robot.chain);
    KDL::Jacobian                   jac(n), jac_push(n);
    KDL::JntArray                   q(n), q_ref(n);
    const KDL::Vector               push_dir(1.0, 0.0, 0.0);
    const KDL::Vector               base(0.0, 0.0, ex::kTableZ);
    Metrics                         m;

    bool restart = false;
    env.on_reset = [&](mj_kdl::ResetContext *ctx) {
        mj_kdl::set_joint_pos(&robot, wp.home);
        ex::place_cube(env);
        ctx->data->ctrl[fingers->ctrl_id] = 0.0;
        ex::prime_gravity(robot, dyn, wp.home);
        m       = {};
        restart = true;
    };
    mj_kdl::reset(&env);
    if (!headless && !mj_kdl::open_viewer(&env)) return 1;

    const auto control = [&](std::size_t, const KDL::JntArray &q_des) {
        q_ref = q_des;
        ex::pd_gravity(robot, dyn, q_des, ex::kKp, ex::kKd);
    };
    // Measured against the reference of the same cycle; the push lands at the next update().
    const auto after_step = [&](const ex::Phase &phase, double t_rel) {
        KDL::Frame elbow;
        mj_kdl::get_body_frame(&env, "forearm_link", &elbow);
        m.elbow_min               = std::min(m.elbow_min, elbow.p.z());
        const bool        pushing = std::string(phase.name) == kPushPhase;
        const KDL::Vector force   = push_dir * (pushing ? kPushForce * push_scale(t_rel) : 0.0);
        push->wrench              = KDL::Wrench(force, KDL::Vector::Zero());
        if (!pushing) return;

        KDL::Frame tcp, ref;
        ex::read_q(robot, q);
        fk.JntToCart(q, tcp);
        fk.JntToCart(q_ref, ref);
        m.cube_slip           = std::max(m.cube_slip, (cube->pose.p - base - tcp.p).Norm());
        const KDL::Vector lag = tcp.p - ref.p;
        if (t_rel < kPushOn) m.lag0 = lag;
        const double deflection = KDL::dot(lag - m.lag0, push_dir);
        if (t_rel >= kPushOff + kSettle && m.residual < 0.0) m.residual = std::abs(deflection);
        // Only at full force: on the ramps the arm lags the changing force.
        const bool full = t_rel >= kPushOn + kPushRamp && t_rel <= kPushOff - kPushRamp;
        if (!full || deflection <= m.peak) return;

        // dq = K^-1 J_push^T F at the pushed body's centre of mass, seen at the TCP through J.
        jac_solver.JntToJac(q_ref, jac);
        const double *com = env.data->xipos + 3 * push->body_id;
        jac_push          = jac;
        jac_push.changeRefPoint(KDL::Vector(com[0], com[1], com[2]) - base - ref.p);
        KDL::Vector dx = KDL::Vector::Zero();
        for (unsigned i = 0; i < n; ++i)
            dx += jac.getColumn(i).vel * (KDL::dot(jac_push.getColumn(i).vel, force) / ex::kKp[i]);
        m.peak      = deflection;
        m.predicted = KDL::dot(dx, push_dir);
    };
    const bool completed = ex::run_phases(
      env, { { &robot, fingers } }, ex::pick_place_phases(wp), restart, control, after_step
    );
    mj_kdl::update(&env);

    const KDL::Vector c        = cube->pose.p;
    const double      place_xy = std::hypot(c.x() - ex::kPlaceXY[0], c.y() - ex::kPlaceXY[1]);
    const bool        on_table = std::abs(c.z() - (ex::kTableZ + ex::kCubeHS)) < 0.002;
    const double      ratio    = m.predicted > 0.0 ? m.peak / m.predicted : 0.0;
    const double      elbow_h  = m.elbow_min - ex::kTableZ;
    std::cout << std::fixed << std::setprecision(4) << "cube final position: [" << c.x() << ", "
              << c.y() << ", " << c.z() << "] place error " << place_xy * 1000.0 << " mm (limit "
              << kMaxPlaceErr * 1000.0 << " mm)" << (on_table ? "" : ", not on the table") << "\n"
              << "push deflection: " << m.peak * 1000.0 << " mm (at least "
              << kMinDeflection * 1000.0 << " mm), stiffness predicts " << m.predicted * 1000.0
              << " mm, ratio " << ratio << " (limits " << kDeflectionRatio[0] << ".."
              << kDeflectionRatio[1] << ")\n"
              << "deflection " << kSettle << " s after the push: " << m.residual * 1000.0
              << " mm (limit " << kMaxResidual * 1000.0 << " mm)\n"
              << "cube from the TCP while pushed: " << m.cube_slip * 1000.0 << " mm (limit "
              << kMaxCubeSlip * 1000.0 << " mm)\n"
              << "lowest elbow above the table: " << elbow_h * 1000.0 << " mm (limit "
              << kMinElbowHeight * 1000.0 << " mm)\n";
    mj_kdl::cleanup(&env);
    const bool ok = completed && on_table && place_xy <= kMaxPlaceErr && m.peak >= kMinDeflection
                    && ratio >= kDeflectionRatio[0] && ratio <= kDeflectionRatio[1]
                    && m.residual >= 0.0 && m.residual <= kMaxResidual
                    && m.cube_slip <= kMaxCubeSlip && elbow_h >= kMinElbowHeight;
    return headless ? ex::verdict(ok, "the cube was placed through the push") : 0;
}

File: python/mj_kdl_wrapper/examples/ex_table_pick_place.py
#!/usr/bin/env python3
"""Table pick-place with joint impedance, pushed while carrying the cube; Python counterpart
of ex_table_pick_place.cpp. Headless it exits 1 unless the cube is placed, the push deflection
matches J K^-1 J^T F, the arm springs back, the cube stays in hand and the elbow stays up."""

from __future__ import annotations

import argparse
import math
from dataclasses import dataclass

import PyKDL as kdl

import mj_kdl_wrapper as mjk

HOME = [0.0, 0.2618, 3.1416, -2.2689, 0.0, 0.9599, 1.5708]
SURFACE_Z = 0.70
CUBE_HS = 0.02
PICK = (0.40, 0.00)
PLACE = (0.40, 0.24)
KP = [100.0, 200.0, 100.0, 200.0, 100.0, 200.0, 100.0]
KD = [10.0, 20.0, 10.0, 20.0, 10.0, 20.0, 10.0]
GRIPPER_CLOSED = 0.82  # [rad] top of the 2F-85's ctrlrange (its driver joint stops at 0.8)
CUBE_START = [PICK[0], PICK[1], SURFACE_Z + CUBE_HS]
MAX_PLACE_ERR = 0.005  # [m] in the table plane

PUSH_BODY = "bracelet_link"
PUSH_PHASE = "PLACE_ABOVE"
PUSH_FORCE = 20.0  # [N] along world +x, across the carry along +y
PUSH_DIR = kdl.Vector(1.0, 0.0, 0.0)
PUSH_ON, PUSH_OFF, PUSH_RAMP = 0.8, 2.0, 0.2  # [s] into the phase; rise and fall
SETTLE = 0.5  # [s] after the push, when the residual is taken

MIN_DEFLECTION = 0.010  # [m]
DEFLECTION_RATIO = (0.85, 1.15)  # measured / predicted
MAX_RESIDUAL = 0.004  # [m]
MAX_CUBE_SLIP = 0.010  # [m] cube centre from the TCP while pushed
MIN_ELBOW_HEIGHT = 0.45  # [m] forearm_link origin above the table


class ResetRequested(Exception):
    """Raised after a simulate UI reset, to restart the sequence."""


@dataclass(frozen=True)
class Phase:
    name: str
    target: list[float]
    duration: float
    timeout: float
    settle_tol: float
    gripper: float


def build_env() -> tuple[mjk.Env, mjk.Robot]:
    table = mjk.SceneObject()
    table.name = "table"
    table.mjcf_path = mjk.menagerie.asset_path("table.xml", env_var="MJ_KDL_TABLE")
    table.pos = [0.0, 0.0, SURFACE_Z]
    table.fixed = True
    cube = mjk.SceneObject()
    cube.name = "cube"
    cube.shape = mjk.Shape.BOX
    cube.size = [CUBE_HS, CUBE_HS, CUBE_HS]
    cube.pos = CUBE_START[:]
    cube.rgba = [0.1, 0.35, 1.0, 1.0]
    cube.mass = 0.1
    cube.condim = mjk.Condim.Torsional
    cube.friction = [0.8, 0.02, 0.001]
    gripper = mjk.AttachmentSpec()
    gripper.mjcf_path = mjk.menagerie.asset_path("robotiq_2f85/2f85.xml", env_var="MJ_KDL_GRIPPER")
    gripper.attach_to = mjk.AttachTarget(mjk.AttachKind.Site, "pinch_site")
    gripper.prefix = "g_"
    robot_spec = mjk.RobotSpec()
    robot_spec.path = mjk.menagerie.model_path("kinova_gen3", env_var="MJ_KDL_MODEL")
    robot_spec.pos = [0.0, 0.0, SURFACE_Z]
    robot_spec.attachments = [gripper]

    spec = mjk.SceneSpec()
    spec.timestep = 0.002
    spec.add_floor = True
    spec.add_skybox = True
    spec.objects = [table, cube]
    spec.robots = [robot_spec]
    env = mjk.Env.build(spec)
    tool = mjk.ToolFrameSpec()
    tool.tool_body = "g_base_mount"
    tool.tcp_site = "g_pinch"
    return env, env.create_robot("base_link", "bracelet_link", tool=tool)


def jnt(values) -> kdl.JntArray:
    q = kdl.JntArray(len(values))
    for i, value in enumerate(values):
        q[i] = value
    return q


def gravity(dyn: kdl.ChainDynParam, q_values) -> list[float]:
    g = kdl.JntArray(len(q_values))
    dyn.JntToGravity(jnt(q_values), g)
    return [g[i] for i in range(g.rows())]


def solve_near_seed(chain, limits, seed: list[float], target: kdl.Frame) -> list[float]:
    """Damped-least-squares IK stepped from seed, so the answer stays on the seed's branch."""
    fk = kdl.ChainFkSolverPos_recursive(chain)
    ik = kdl.ChainIkSolverVel_wdls(chain, 1e-5, 150)
    ik.setLambda(0.05)
    q = jnt(seed)
    dq = kdl.JntArray(q.rows())
    for _ in range(300):
        current = kdl.Frame()
        fk.JntToCart(q, current)
        dx = kdl.diff(current, target)
        if dx.vel.Norm() <= 2e-3 and dx.rot.Norm() <= 2e-2:
            break
        if dx.vel.Norm() > 0.05:
            dx.vel = dx.vel * (0.05 / dx.vel.Norm())
        if dx.rot.Norm() > 0.20:
            dx.rot = dx.rot * (0.20 / dx.rot.Norm())
        if ik.CartToJnt(q, dx, dq) < 0:
            raise RuntimeError("PyKDL IK velocity step failed")
        # joint_limits are +-inf for an unlimited joint, so the clamp leaves it alone.
        for i in range(q.rows()):
            q[i] = min(limits[i][1], max(limits[i][0], q[i] + dq[i]))
    else:
        current = kdl.Frame()
        fk.JntToCart(q, current)
        dx = kdl.diff(current, target)
        if dx.vel.Norm() > 2e-3 or dx.rot.Norm() > 2e-2:
            raise RuntimeError("PyKDL IK did not converge")
    return [q[i] for i in range(q.rows())]


def pick_place_phases(robot: mjk.Robot) -> list[Phase]:
    chain = robot.kdl_chain()
    seed = HOME[:]

    # Base-frame heights: the TCP at the cube centre, then above it.
    def solve(xy, z: float) -> list[float]:
        nonlocal seed
        target = kdl.Frame(robot.tip_T_tcp.M, kdl.Vector(xy[0], xy[1], z))
        seed = solve_near_seed(chain, robot.joint_limits, seed, target)
        return seed[:]

    pick_above = solve(PICK, CUBE_HS + 0.20)
    pick = solve(PICK, CUBE_HS)
    lift = solve(PICK, CUBE_HS + 0.30)
    place_above = solve(PLACE, CUBE_HS + 0.20)
    place = solve(PLACE, CUBE_HS)
    closed = GRIPPER_CLOSED
    return [
        Phase("HOME", HOME, 1.0, 2.5, 0.08, 0.0),
        Phase("PICK_ABOVE", pick_above, 5.0, 7.0, 0.08, 0.0),
        Phase("PICK", pick, 5.0, 8.0, 0.03, 0.0),
        Phase("CLOSE", pick, 1.5, 2.5, -1.0, closed),
        Phase("LIFT", lift, 3.0, 5.0, 0.08, closed),
        Phase("PLACE_ABOVE", place_above, 3.0, 5.0, 0.08, closed),
        Phase("PLACE", place, 5.0, 8.0, 0.03, closed),
        Phase("OPEN", place, 1.0, 2.0, -1.0, 0.0),
        Phase("RETREAT", place_above, 2.0, 4.0, 0.08, 0.0),
        Phase("HOLD", place_above, 1.0, 1.0, -1.0, 0.0),
    ]


def new_metrics() -> dict:
    return {
        "lag0": kdl.Vector(),  # TCP minus its reference just before the push
        "peak": 0.0,  # [m] along the push
        "predicted": 0.0,  # [m] J K^-1 J^T F at the peak
        "residual": -1.0,  # [m] along the push, SETTLE after it; < 0 until taken
        "cube_slip": 0.0,
        "elbow_min": math.inf,  # [m] world z
    }


def push_scale(t_rel: float) -> float:
    rise = (t_rel - PUSH_ON) / PUSH_RAMP
    fall = (PUSH_OFF - t_rel) / PUSH_RAMP
    return min(1.0, max(0.0, min(rise, fall)))


def measure_push(env, robot, state: dict, q_ref, cube, t_rel: float, force) -> None:
    """Deflection against the reference of the same cycle, and what the stiffness predicts."""
    m, fk, jac_solver = state["metrics"], state["fk"], state["jac"]
    base = kdl.Vector(0.0, 0.0, SURFACE_Z)
    tcp, ref = kdl.Frame(), kdl.Frame()
    fk.JntToCart(jnt(robot.jnt_pos_msr), tcp)
    fk.JntToCart(jnt(q_ref), ref)
    m["cube_slip"] = max(m["cube_slip"], (cube - base - tcp.p).Norm())
    lag = tcp.p - ref.p
    if t_rel < PUSH_ON:
        m["lag0"] = lag
    deflection = kdl.dot(lag - m["lag0"], PUSH_DIR)
    if t_rel >= PUSH_OFF + SETTLE and m["residual"] < 0.0:
        m["residual"] = abs(deflection)
    # Only at full force: on the ramps the arm lags the changing force.
    full = PUSH_ON + PUSH_RAMP <= t_rel <= PUSH_OFF - PUSH_RAMP
    if not full or deflection <= m["peak"]:
        return

    # dq = K^-1 J_push^T F at the pushed body's centre of mass, seen at the TCP through J.
    n = robot.n_joints
    jac = kdl.Jacobian(n)
    jac_solver.JntToJac(jnt(q_ref), jac)
    jac_push = kdl.Jacobian(jac)
    jac_push.changeRefPoint(kdl.Vector(*env.data.body(PUSH_BODY).xipos) - base - ref.p)
    dx = kdl.Vector()
    for i in range(n):
        dx += jac.getColumn(i).vel * (kdl.dot(jac_push.getColumn(i).vel, force) / KP[i])
    m["peak"] = deflection
    m["predicted"] = kdl.dot(dx, PUSH_DIR)


def run_phase(env, robot, fingers, dyn, phase: Phase, state: dict) -> bool:
    """One update() per step: it reads the state and applies the previous cycle's command."""
    print(f"State: {phase.name}")
    start = robot.jnt_pos_msr[:]
    t0 = env.data.time
    push = env.data.body(PUSH_BODY)
    pushing = phase.name == PUSH_PHASE
    while True:
        env.update()
        elapsed = env.data.time - t0
        alpha = min(1.0, max(0.0, elapsed / phase.duration)) if phase.duration > 0.0 else 1.0
        q, qd = robot.jnt_pos_msr, robot.jnt_vel_msr
        q_ref = [start[i] + alpha * (phase.target[i] - start[i]) for i in range(len(q))]
        g = gravity(dyn, q)
        robot.jnt_trq_cmd = [
            g[i] + KP[i] * (q_ref[i] - q[i]) - KD[i] * qd[i] for i in range(len(q))
        ]
        fingers.ctrl[0] = phase.gripper
        cube = env.body_frame("cube").p if pushing else None
        err = max(abs(t - m) for t, m in zip(phase.target, q))
        done_pose = phase.settle_tol < 0.0 or err <= phase.settle_tol
        if (elapsed >= phase.duration and done_pose) or elapsed >= phase.timeout:
            return True
        if not env.step():
            return False
        if state["reset"]:
            state["reset"] = False
            raise ResetRequested()
        env.pace()
        metrics = state["metrics"]
        metrics["elbow_min"] = min(metrics["elbow_min"], env.body_frame("forearm_link").p.z())
        force = PUSH_DIR * (PUSH_FORCE * push_scale(elapsed) if pushing else 0.0)
        push.xfrc_applied[:] = [force.x(), force.y(), force.z(), 0.0, 0.0, 0.0]
        if pushing:
            measure_push(env, robot, state, q_ref, cube, elapsed, force)


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--gui", action="store_true")
    args = parser.parse_args()

    env, robot = build_env()
    try:
        robot.set_control_mode(mjk.CtrlMode.TORQUE)
        fingers = env.data.actuator("g_fingers_actuator")
        chain = robot.kdl_chain()
        dyn = kdl.ChainDynParam(chain, kdl.Vector(0.0, 0.0, env.spec.gravity_z))
        phases = pick_place_phases(robot)
        state = {
            "reset": False,
            "metrics": new_metrics(),
            "fk": kdl.ChainFkSolverPos_recursive(chain),
            "jac": kdl.ChainJntToJacSolver(chain),
        }

        def on_reset(ctx):
            robot.set_joint_pos(HOME)
            env.set_body_pose("cube", CUBE_START)
            fingers.ctrl[0] = 0.0
            robot.jnt_trq_cmd = gravity(dyn, HOME)
            state["metrics"] = new_metrics()
            state["reset"] = True

        env.on_reset = on_reset
        env.reset()
        state["reset"] = False
        if args.gui:
            env.open_viewer("ex_table_pick_place.py")

        completed = False
        while not completed:
            try:
                completed = all(run_phase(env, robot, fingers, dyn, p, state) for p in phases)
                break
            except ResetRequested:
                continue
        env.update()

        cube = env.body_frame("cube").p
        place_err = math.hypot(cube.x() - PLACE[0], cube.y() - PLACE[1])
        on_table = abs(cube.z() - (SURFACE_Z + CUBE_HS)) < 0.002
        m = state["metrics"]
        ratio = m["peak"] / m["predicted"] if m["predicted"] > 0.0 else 0.0
        elbow_h = m["elbow_min"] - SURFACE_Z
        print(
            f"cube final position: [{cube.x():.4f}, {cube.y():.4f}, {cube.z():.4f}] "
            f"place error {place_err * 1000:.4f} mm (limit {MAX_PLACE_ERR * 1000} mm)"
            + ("" if on_table else ", not on the table")
        )
        print(
            f"push deflection: {m['peak'] * 1000:.4f} mm (at least {MIN_DEFLECTION * 1000} mm), "
            f"stiffness predicts {m['predicted'] * 1000:.4f} mm, ratio {ratio:.4f} "
            f"(limits {DEFLECTION_RATIO[0]}..{DEFLECTION_RATIO[1]})"
        )
        print(
            f"deflection {SETTLE} s after the push: {m['residual'] * 1000:.4f} mm "
            f"(limit {MAX_RESIDUAL * 1000} mm)"
        )
        print(
            f"cube from the TCP while pushed: {m['cube_slip'] * 1000:.4f} mm "
            f"(limit {MAX_CUBE_SLIP * 1000} mm)"
        )
        print(
            f"lowest elbow above the table: {elbow_h * 1000:.4f} mm "
            f"(limit {MIN_ELBOW_HEIGHT * 1000} mm)"
        )
    finally:
        env.close()
    if args.gui:
        return 0
    ok = (
        completed
        and on_table
        and place_err <= MAX_PLACE_ERR
        and m["peak"] >= MIN_DEFLECTION
        and DEFLECTION_RATIO[0] <= ratio <= DEFLECTION_RATIO[1]
        and 0.0 <= m["residual"] <= MAX_RESIDUAL
        and m["cube_slip"] <= MAX_CUBE_SLIP
        and elbow_h >= MIN_ELBOW_HEIGHT
    )
    print(f"{'PASS' if ok else 'FAIL'}: the cube was placed through the push")
    return 0 if ok else 1


if __name__ == "__main__":
    raise SystemExit(main())
